man/man1/fltinv.1
.TH FLTINV 1 "February 2020" "Version 2020.02.01" "User Manuals"

.SH NAME
fltinv \- invert displacement or pre-stresses for fault slip.

.SH SYNOPSIS
.P
.B fltinv
.BI -mode " inversion_mode"
.BI -o " output_file"
.BI "-flt " fault_file
.BI [-flt:rake " rake_constraint_file" ]
.BI [-flt:slip " slip_constraint_file" ]
.BI [-euler " euler_input_file euler_output_file" ]
.BI -disp " displacement_file"
.B |
.BI -los " los_file"
.B |
.BI -prests " prestress_file"
.BI "[-disp:components [" 1 ][ 2 ][ 3 ]]
.BI [-cov " covariance_file" ]
.BI [-disp:unit " input_unit" ]
.BI [-disp:misfit " misfit_file" ]
.BI -gf:model " model"
.BI [-gf:disp_file " gf_disp_file" ]
.BI [-gf:los_file " gf_los_file" ]
.BI [-gf:stress_file " gf_stress_file" ]
.B -xy|-geo
.BI [-haf " halfspace_file"]
.BI [-damp " damping_constant"]
.BI [-smoothing " smoothing_constant smoothing_file"]
.BI [-lsqr:mode " gels" | nnls | gesv ]
.BI [-anneal:init_mode " mean" | min | max | rand | user | unlocked | locked ]
.BI [-anneal:step " step_size" ]
.BI [-anneal:max_it " max_iteration" ]
.BI [-anneal:reset_iteration " reset_iteration" ]
.BI [-anneal:temp_start " temp_start" ]
.BI [-anneal:temp_min " temp_minimum" ]
.BI [-anneal:cool " cooling_factor" ]
.BI [-anneal:log_file " anneal_log_file" ]
.BI [-anneal:seed " seed" ]
.B [-anneal:model_uncertainty]
.BI [-anneal-psc:min_flip " nmin" ]
.BI [-anneal-psc:max_flip " nmax" ]
.BI [-v " level" ]


.SH DESCRIPTION
.B fltinv
inverts for fault slip distributions based on either displacement observations
or pre-stresses resolved onto the faults. The inversion can be done using linear least
squares, simulated annealing, or the neighborhood algorithm (j/k, not yet). The meaning of
the inputs and the specific outputs depend on the mode and other control specifications.


.SH OPTIONS

.TP
.BI -mode " inversion_mode"
Specify the
.I inversion_mode
as one of:
.BR lsqr " (default), " anneal ", or " anneal-psc .

.BR lsqr :
Linear least-squares inversion. Loads Green's functions, regularization, and constraints
into model matrix and into observation vector. Uses gels (LAPACK generalized least squares),
nnls (Non-negative least-squares), or gesv (LAPACK generalized square matrix solver) algorithm
to invert for fault slip values.

.BR anneal :
Simulated annealing inversion. Starts with an initial fault solution, then
randomly perturbs fault slip and rake values. If objective function
(chi-squared misfit) decreases, keeps new solution. If objective increases,
keeps new solution with a probability that decreases over the search
(associated with the "temperature" of the annealing
process). For more information, see Sambridge and Mosegaard (2002).

.BR anneal-psc :
Simulated annealing inversion with pseudo-coupling as defined by Herman et al. (2018).
Faults are either completely locked or unlocked, then randomly switched in each iteration.
Computes partial slip around locked sections using least squares algorithm gesv (LAPACK solver).
Computes predicted displacement and corresponding misfits to displacement observations.

.TP
.B OUTPUT OPTIONS

.TP
.BI -o " output_file"
The inverted fault slip solution has the format:
.IR "strike-slip dip-slip" ,
for each fault in the input
.IR fault_file ,
including faults with fixed slip values.


.TP
.B FAULT OPTIONS

.TP
.BI "-flt " fault_file
File containing parameters of faults for inversion. The file format depends on the
Green's function calculation model type, specified using option
.BR -gf:model :

.BR okada_rect :
.I x y z str dip wid len
.br
.BR okada_pt :
.I x y z str dip area
.br
.BR triangle :
.I x1 y1 z1 x2 y2 z2 x3 y3 z3

The
.I fault_file
needs to be in its full format if GFs must be computed for either displacements or stresses.
All values are in SI units. If all desired GFs are pre-computed (options
.BR -gf:disp_file ", " -gf:los_file ", or " -gf:sts_file ),
this file can contain anything as long as the number of lines corresponds to the correct
number of faults required for the slip degrees of freedom defined by the GFs.


.TP
.BI "-flt:rake " rake_constraint_file
Constraints on the rake angles. If
.I rake_constraint_file
has only one line, use that value for all faults. Otherwise, the rake angles in the file
must correspond to the faults in
.IR fault_file .
The meaning of the values depends on the
.IR inversion_mode
and the number of columns:

.BR lsqr " (one column):"
fix the rake of each fault to the corresponding value in
.I rake_constraint_file

.BR lsqr " (two columns):"
set the two rakes for calculating Green's functions to the corresponding values in
.I rake_constraint_file
instead of 0 and 90 (this is useful for constraining rake to one quadrant in conjunction with
.BR nnls )

.BR anneal " (two columns):"
define the range of possible rake angles for each fault in the search

.BR anneal-psc " (one column):"
fix the rake of each fault to the corresponding value in
.I rake_constraint_file


.TP
.BI "-flt:slip " slip_constraint_file
Constraints on the slip magnitudes. If
.I slip_constraint_file
has only one line, use that value for all faults. Otherwise, slip magnitudes in the file
must correspond to the faults in
.IR fault_file .
Meaning of values in file depends on value of
.IR inversion_mode :

.B lsqr
.IR "" ( "strike_slip dip_slip" ):
set fault slip to fixed values specified in
.IR slip_constraint_file ;
set to 99999 to NOT fix the value

.B anneal
.IR "" ( "slip_min slip_max" ):
minimum and maximum slip magnitudes in each fault

.B anneal-psc
.IR "" ( "strike_slip dip_slip" ):
slip magnitudes for locked faults; set to 99999 to keep fault always unlocked



.TP
.B RIGID ROTATION (EULER POLE) OPTIONS

.TP
.BI -euler " euler_input_file euler_output_file"
Invert for (multiple) rigid body rotations, with output in format:
.I pole_lon pole_lat
.IR ang_vel (deg/Ma).
The file
.I euler_input_file
has the format:

.I n
.br
.IR "" [ "lon_1 lat_1 rad_1 min_rate_1 max_rate_1" ]
.br
.IR "" [ "lon_2 lat_2 rad_2 min_rate_2 max_rate_2" ]
.br
 ...
.br
.IR "" [ "lon_n lat_n rad_n min_rate_n max_rate_n" ]
.br
.IR "ipole 3" | "L  iobs"
 ...

where
.I n
is the number of Euler poles to solve for, and
.IR lon_i ,
.IR lat_i ,
.IR rad_i (km),
.IR min_rate_i (deg/Ma),
.IR max_rate_i (deg/Ma),
define the range for the possible Euler poles in annealing searches
(these lines are not used and can be left blank in
.B lsqr
mode).
The last data block defines which points correspond to which rigid rotation pole.
.I i_pole
is the pole for the point,
.IR 3 | L
indicates whether the observation is three-component or line-of-sight,
and
.I iobs
is the index of the observation corresponding to
.I displacement_file
or
.IR los_file .


.TP
.B INPUT OPTIONS

.TP
.BI -disp " displacement_file"
Three-component displacement observations in the format:
.IR "x y z ux uy uz" .
The misfit between these displacements and those produced by modeled fault slip will be minimized.
All values are in SI units, unless the coordinate type is changed to geographic using the
.B -geo
option.

.TP
.BI "-disp:components [" 1 ][ 2 ][ 3 ]
Use only selected displacement components in the inversion. Default is
.IR 123 .
Useful when only horizontal or vertical data are available or in 2D problems.

.TP
.BI -los " los_file"
Line-of-sight displacement observations in the format:
.IR "x y z ulos azimuth inclination" .
.I Azimuth
is the horizontal look direction from the observer to the site, clockwise from north.
.I Inclination
is the vertical look direction from observer to site, from 0 at the horizon to 90 at vertical
(i.e., the observer is directly above the site).
The misfit between these displacements and those produced by modeled fault slip will be minimized.
All values are in SI units, unless the coordinate type is changed to geographic using the
.B -geo
option, and angles are in degrees.

.TP
.BI -prests " prestress_file"
Pre-stresses on all input faults in the format:
.IR "sxx syy szz sxy sxz syz" .
The shear stresses resolved onto the faults will be minimized. Note that
the pre-stress values can be set to zero to minimize shear stresses without
external influences; this is particularly useful in combination with fixing
the fault slip values to provide a source of stress. In the case where only pre-stress values
are provided (plus optional fixed-slip faults) without kinematic
observations, the matrix equation is solved directly using the
LAPACK routine dgesv instead of with a least-squares approach, because the system
of equations is properly determined and well behaved. All values are in SI units.

.TP
.BI -cov " covariance_file"
File with displacement covariance data in the format:
.IR "idisp jdisp icmp jcmp cov" ,
where
.I idisp
and
.I jdisp
are the indices of the displacement corresponding to the entries in
.IR displacement_file ,
.I icmp
and
.I jcmp
are the components of the displacement or L for line-of-sight displacement,
and
.I cov
is the covariance, in square meters.


.TP
.BI -disp:unit " input_unit"
Define the units of input displacement and los files.
Can also redefine them as velocities here (yeah, I know the option name is "disp";
deal with it). Options:
.I m
(meters, default),
.I mm
(millimeters),
.I m/s
(meters/second),
.I m/yr
(meters/year),
.I mm/s
(millimeters/second),
.I mm/yr
(millimeters/year).

.TP
.B -xy
Treat
.I x
and
.I y
as SI units in all input files

.TP
.B -geo
Treat
.I x
and
.I y
as geographic coordinates instead of SI units in all input files




.TP
.B OUTPUT OPTIONS
.TP
.BI -disp:misfit " misfit_file"
Calculate the root-mean-squared misfit to the observations.


.TP
.B GREENS FUNCTIONS OPTIONS
.TP
.BI -gf:model " model"
Model to be used to calculate Green's functions:
.BR okada_rect ", " okada_pt ", or " triangle .

.TP
.BI -gf:disp_file " gf_disp_file"
Pre-computed three-component displacement Green's functions, e.g., from a finite element model.
All values are in SI units.
The file structure is as follows:

       1 - m  |  1 - m
.br
     _________|_________
.br
    [         |         ]
.br
  1 [         |         ]
.br
  | [  ux_ss  |  ux_ds  ]
.br
  n [         |         ]
.br
 ___[_________|_________]
.br
    [         |         ]
.br
  1 [         |         ]
.br
  | [  uy_ss  |  uy_ds  ]
.br
  n [         |         ]
.br
 ___[_________|_________]
.br
    [         |         ]
.br
  1 [         |         ]
.br
  | [  uz_ss  |  uz_ds  ]
.br
  n [         |         ]
.br
    [_________|_________]

where n is the number of displacement observations, m is the number of faults,
ux, uy, uz are the components of displacement, and ss and ds are strike-slip
and dip-slip fault sources.

.TP
.BI -gf:los_file " gf_los_file"
Pre-computed line-of-sight displacement Green's functions, e.g., from a finite element model.
All values are in SI units.
The file structure is as follows:

       1 - m  |  1 - m
.br
     _________|_________
.br
    [         |         ]
.br
  1 [         |         ]
.br
  | [  uL_ss  |  uL_ds  ]
.br
  n [         |         ]
.br
    [_________|_________]

where n is the number of displacement observations, m is the number of faults,
uL is the line-of-sight displacement, and ss and ds are strike-slip
and dip-slip fault sources.

.TP
.BI -gf:stress_file " gf_stress_file"
Pre-computed stress Green's functions, e.g., from a finite element model.
All values are in SI units.
The file structure is as follows:

       1 - m  |  1 - m
.br
     _________|_________
.br
    [         |         ]
.br
  1 [ ss sts  | ss sts  ]
.br
  | [    /    |    /    ]
.br
  m [  ss slp |  ds slp ]
.br
 ___[_________|_________]
.br
    [         |         ]
.br
  1 [ ds sts  | ds sts  ]
.br
  | [    /    |    /    ]
.br
  m [  ss slp |  ds slp ]
.br
    [_________|_________]

where m is the number of faults, ss sts and ds sts are the strike-slip and dip-slip
components of the shear traction resolved onto the fault plane, and ss slp and ds slp
are strike-slip and dip-slip fault sources.




.TP
.B HALF-SPACE OPTIONS

.TP
.BI -haf " halfspace_file"
Optional file to define half-space elastic moduli. The default values
are
.IR lame "=40e9 Pa and " shearmod "=40e9 Pa."
This file has the following format:
.IR "modulus1 value1 modulus2 value2" ,
where
.I modulus
can be "lame", "shearmod", "poisson", or "young" and
.I value
is the value in Pa.



.TP
.B REGULARIZATION OPTIONS

.TP
.BI "-damp " damping_constant
Minimize the L1 length of the fault slip solution with a weighting factor of
.IR damping_constant * damping_constant

.TP
.BI "-smoothing " "smoothing_constant smoothing_file"
Minimize the Laplacian roughness of the fault slip solution with a weighting factor of
.IR smoothing_constant * smoothing_constant .
The file
.I smoothing_file
describes the connectivity between fault segments and has the following format:

ifault nneighbors neighbor_1 neighbor_2 ... neighbor_n







.TP
.B LEAST SQUARES OPTIONS

.TP
.BI "-lsqr:mode " gels | nnls | gesv
Set the algorithm to use for the least-squares solution (default: gels).
The
.I gels
and
.I gesv
algorithms are from the LAPACK library.
The
.I nnls
algorithm is from Lawson and Hanson (1974), translated into Fortran 90
by Alan Miller.





.TP
.B SIMULATED ANNEALING OPTIONS

.P
The simulated annealing algorithm is a combined parameter exploration and optimization technique
(Kirkpatrick et al., 1983). The concept is straightforward; start with an initial model (combination
of parameter values) and set it to be the "current" model. In each iteration, a new model is
proposed by perturbing the current model. If the "proposed" model results in an improved objective
function (f, e.g., lower data misfit), then it is accepted as the current model. If the proposed
model results in a worse objective function, then it can still be accepted with a probability
proportional to how much worse the objective function has gotten:
.br
             f(i)-f(i-1)
.br
    p = exp( ----------- )
.br
                  T
.br
where T is a "temperature" term that controls the acceptance rate. At high T, many worse fitting
models will be accepted, encouraging exploration of the parameter space. At low T, only improved
models will be accepted, moving towards the optimum of the objective function. The algorithm
systematically progresses from high T to low T.

.P
Model parameter statistics can be derived using the Metropolis-Hastings algorithm (Metropolis et
al., 1953; Hastings, 1970), which is simply the simulated annealing algorithm setting T=1.
It is common to use the chi-squared misfit for the objective function. Thus, setting the annealing
parameters carefully, it is possible to estimate parameters statistics.

.TP
.BI "-anneal:init_mode " mean | min | max | rand | user | unlocked | locked
Set the mode to initialize the slip magnitudes and rake angles for mode
.B anneal
or to initialize which faults are locked and unlocked for mode
.BR anneal-psc .

.BR anneal :
.br
.IR mean :
set values to center of slip and rake ranges
.br
.IR min :
set values to minimum values of slip and rake ranges
.br
.IR max :
set values to maximum values of slip and rake ranges
.br
.IR rand :
randomly select starting value from uniform distribution in ranges
.br
.IR user :
specify a file with starting slip and rake

.BR anneal-psc :
.IR unlocked :
set all faults to start unlocked
.br
.IR locked :
set all faults to start locked
.br
.IR randODDS :
randomly set locked faults with probability, e.g., rand0.75 for likelihood a fault is
locked of 0.75
.br
.IR user :
specify a file with starting locked (1) and unlocked (0) faults

.TP
.BI -anneal:step " step_file"
Slip magnitude and rake angle step sizes for each fault in annealing search.

.TP
.BI -anneal:max_it " max_iteration"
Set the maximum number of iterations in the simulated annealing search
(default: 1000).

.TP
.BI -anneal:reset_it " reset_iteration"
After each
.I reset_iteration
steps, the temperature in the annealing search will be set to
.I temp_start
(default: 1000000).

.TP
.BI -anneal:temp_start " temp_start"
Starting temperature for annealing process (default: 2.0).

.TP
.BI -anneal:temp_min " temp_minimum"
Minimum temperature for annealing process (default: 0.0).

.TP
.BI -anneal:cool " cooling_factor"
Factor to reduce temperature by every iteration (default: 0.98).

.TP
.BI -anneal:log_file " anneal_log_file"
Save a log of solutions and fits in
.IR anneal_log_file .

.TP
.BI -anneal:seed " seed"
Set the random number generator seed (default: based on date and time)

.TP
.B -anneal:model_uncertainty
Search for the model prediction error in addition to the model parameters.
The model prediction error is presented as a fraction of the observation
magnitude (Minson et al., 2013).

.TP
.BI -anneal-psc:min_flip " nmin"
Minimum number of faults to flip from locked to unlocked or vice versa every iteration

.TP
.BI -anneal-psc:max_flip " nmax"
Maximum number of faults to flip from locked to unlocked or vice versa every iteration




.TP
.B MISCELLANEOUS OPTIONS

.TP
.BI "-v " level
Set the verbosity of the program. Level 1 prints major steps, level 2 prints subroutine
start/finish, and level 3 prints all intermediate calculations.




.SH REFERENCES
Hastings, W.K. (1970).
Monte carlo sampling methods using Markov chains and their applications.
Biometrika 57.

Herman, M.W., Furlong, K.P., Govers, R. (2018).
The accumulation of slip deficit in subduction zones in the absence of mechanical
coupling: Implications for the behavior of megathrust earthquakes.
Jouranl of Geophysical Research 123.

Kirkpatrick, S., Gelatt, C.D., Vecchi, M.P. (1983).
Optimization by simulated annealing.
Science 220.

Lawson, C.L., Hanson, R.J. (1995). Solving Least Squares Problems. SIAM.

Metropolis, N., Rosenbluth, A.W., Rosenbluth, M.N., Teller, A.H., & Teller, E. (1953).
Equation of State Calculations by Fast Computing Machines.
The Journal of Chemical Physics 21.

Minson, S.E., Simons, M., Beck, J.L. (2013).
Bayesian inversion for finite fault earthquake source models I-theory and algorithm.
Geophysical Journal International 194.

Sambridge, M., Mosegaard K. (2002).
Monte Carlo Methods in Geophysical Inverse Problems.
Reviews of Geophysics 40.
.RS


src/fltinv.f90
!--------------------------------------------------------------------------------------------------!
! FLTINV
!
! Program for inverting geodetic observations (and more) to determine fault slip (and more).
!
! Input datasets:
!     - 3-D displacement/velocity vectors (GPS)
!     - Line-of-sight displacement/velocity (InSAR)
!     - Data/model covariances
!     - Shear tractions on faults
!     - Model parameters/Green's functions
!
! Output quantities:
!     - Fault slip
!     - Locked faults
!     - Euler poles (rigid body rotations)
!     - Data misfit
!
! Inversion algorithms:
!     - Linear least-squares
!     - Simulated annealing
!
! Regularization options:
!     - Damping (minimizing model length)
!     - Smoothing (minimizing model roughness)
!
!--------------------------------------------------------------------------------------------------!

module fltinv
!----
! The fltinv module contains global control and data variables, as well as generic I/O subroutines
! for the program.
!----

! Derived data type containing file name, array size, and array information
type fltinv_data
    character(len=512) :: file
    character(len=16) :: flag
    integer :: nrows
    integer :: ncols
    double precision, allocatable :: array(:,:)
end type fltinv_data


! General program control variables
character(len=512) :: output_file                ! Name of output fault slip file
character(len=16) :: inversion_mode              ! Inversion mode: lsqr, anneal, anneal-psc

! Fault parameters
type(fltinv_data) :: fault                       ! Sub-fault locations and geometries
type(fltinv_data) :: slip_constraint             ! Fault slip magnitude constraints
type(fltinv_data) :: rake_constraint             ! Fault rake angle constraints

! Euler pole parameters
character(len=512) :: euler_file                 !
character(len=512) :: euler_output_file          !
integer :: npoles                                !
integer, allocatable :: rigid_pt_array_disp(:)   !
integer, allocatable :: rigid_pt_array_los(:)    !
double precision, allocatable :: pole_array(:,:) !
double precision, allocatable :: gf_euler(:,:)   !

! Observation constraints on fault slip
type(fltinv_data) :: displacement                ! Three-component displacement observations
character(len=3) :: disp_components              ! Components of 3-D displacement to invert for
character(len=512) :: disp_misfit_file           ! RMS misfit to 3-D displacements
type(fltinv_data) :: los                         ! Line-of-sight displacement observations
type(fltinv_data) :: prestress                   ! Stress field at each sub-fault
character(len=512) :: cov_file                   ! File with data covariance data
double precision, allocatable :: cov_matrix(:,:) ! Data covariance matrix
logical :: isCovMatrixDiagonal                   ! Data covariance matrix
character(len=16) :: input_disp_unit             !

! Green's function variables
character(len=16) :: gf_model                    ! Model to compute GFs: okada_rect, okada_pt, triangle, user
type(fltinv_data) :: gf_disp                     ! Three-component displacement Green's functions
type(fltinv_data) :: gf_los                      ! Line-of-sight displacement Green's functions
type(fltinv_data) :: gf_stress                   ! Shear stress Green's functions
character(len=16) :: coord_type                  ! Fault/observation coordinate type: cartesian, geographic

! Elastic properties for calculating Green's functions
character(len=512) :: halfspace_file             ! File with elastic half-space parameters
double precision :: poisson                      ! Poisson's ratio
double precision :: shearmod                     ! Shear modulus
double precision :: lame                         ! Lame's parameter

! Regularization variables
double precision :: damping_constant             ! Minimize L1 norm of solution with this weight
double precision :: smoothing_constant           ! Minimize Laplacian roughness of solution with this weight
character(len=512) :: smoothing_file             ! Smoothing network of neighboring faults
integer :: nsmooth                               ! Number of faults to smooth
integer, allocatable :: smoothing_pointers(:,:)  ! Array to store smoothing pointers
integer, allocatable :: smoothing_neighbors(:)   ! Array to store smoothing neighbors

! Least squares variables
character(len=16) :: lsqr_mode                   ! Least-squares routine: gels, nnls
double precision, allocatable :: Asave(:,:)      ! Saved model matrix

! ! Annealing control variables
character(len=16) :: anneal_init_mode            ! Starting solution: zero, mean, rand, user
character(len=512) :: anneal_init_file           ! User initialization file
character(len=512) :: anneal_step_file           ! File with fault slip and rake step sizes
double precision, allocatable :: step(:)         ! Step size array
integer :: max_iteration                         ! Number of steps in search
integer :: reset_iteration                       ! Reset temperature every N steps
double precision :: temp_start                   ! Initial temperature
double precision :: temp_minimum                 ! Temperature does not decrease below this value
double precision :: cooling_factor               ! Reduce temp by this factor every iteration
character(len=512) :: anneal_log_file            ! Annealing log file
integer :: anneal_seed                           ! Random number seed for annealing (useful for repeating runs)
integer :: min_flip                              ! Minimum number of faults to flip locked<->unlocked
integer :: max_flip                              ! Maximum number of faults to flip locked<->unlocked
logical :: modelUncertainty                      ! Search for model uncertainty

! Output variables
double precision, allocatable :: fault_slip(:,:)
double precision, allocatable :: euler_pole(:,:)

! double precision :: stress_weight                ! Weight to stress data in inversion

! character(len=512) :: los_misfit_file

!--------------------------------------------------------------------------------------------------!
contains
!--------------------------------------------------------------------------------------------------!

subroutine init_fltinv_data(val)
!----
! Reset derived data type fltinv_data variable values
!----

implicit none

! Arguments
type(fltinv_data) :: val

val%file = 'none'
val%flag = ''
val%nrows = 0
val%ncols = 0

if (allocated(val%array)) then
    deallocate(val%array)
endif

return
end subroutine

!--------------------------------------------------------------------------------------------------!

subroutine read_fltinv_data_file(val,ierr)
!----
! Read derived data type fltinv_data from file and set variable values
!----

use io, only: stderr, stdout, fileExists, line_count, verbosity

implicit none

! Arguments
type(fltinv_data) :: val

! Local variables
integer :: i, j, ios, ierr
character(len=32) :: fmt_str
character(len=512) :: input_line


ierr = 0

! Check that a file should be read
if (val%file.eq.'none') then
    write(stderr,*) 'read_fltinv_data_file: no file specified'
    ierr = 1
    return
endif
if (verbosity.ge.2) then
    write(stdout,*) 'read_fltinv_data_file: reading file "',trim(val%file),'"'
endif


! Check that ncols has been specified
if (val%ncols.le.0) then
    write(stderr,*) 'read_fltinv_data_file: ncols has not been defined'
    ierr = 1
    return
endif


! Check that the file exists, then count the number of lines
if (.not.fileExists(val%file)) then
    write(stderr,*) 'read_fltinv_data_file: no file found named "',trim(val%file),'"'
    ierr = 1
    return
endif
val%nrows = line_count(val%file)


! Allocate memory for the associated array(s)
if (.not.allocated(val%array)) then
    allocate(val%array(val%nrows,val%ncols),stat=ierr)
    if (ierr.ne.0) then
        write(stderr,*) 'read_fltinv_data_file: error allocating memory to array read from file',&
                        trim(val%file)
        return
    endif
endif


! Read the file, in free format
open(unit=21,file=val%file,status='old')
do i = 1,val%nrows
    read(21,'(A)',iostat=ios,err=1002,end=1002) input_line
    read(input_line,*,iostat=ios) (val%array(i,j),j=1,val%ncols)
    1002 if (ios.ne.0) then
        write(stderr,*) 'read_fltinv_data_file: read error on file ',trim(val%file),' at line ',i
        write(stderr,*) 'offending line: ',trim(input_line)
        ierr = 1
        return
    endif
enddo
close(21)


! Print finished message
if (verbosity.ge.2) then
    write(stdout,*) 'read_fltinv_data_file: finished reading file "',trim(val%file),'"'
endif
if (verbosity.ge.3) then
    write(stdout,'(" nrows: ",I5)') val%nrows
    write(fmt_str,1001) val%ncols
    1001 format('(1P',I5,'E14.6)')
    do i = 1,val%nrows
        write(stdout,fmt=fmt_str) (val%array(i,j),j=1,val%ncols)
    enddo
endif

return
end subroutine


end module





!==================================================================================================!
!==================================================================================================!
!==================================== MAIN FLTINV PROGRAM =========================================!
!==================================================================================================!
!==================================================================================================!


program main
!----
! The main fltinv program only contains a few calls to subroutines that do all the heavy lifting.
!----

#ifndef USE_LAPACK
    use io, only: stderr
#endif

use fltinv, only: output_file, &
                  euler_output_file, &
                  inversion_mode

implicit none


! Check whether LAPACK libraries are linked during compiling; fltinv will not do much without them.
#ifndef USE_LAPACK
    write(stderr,*) 'fltinv does not do much compiled without LAPACK libraries'
    write(stderr,*) 'In fact, all it does is print this message and exit with an error'
    call error_exit(1)
#endif


! Parse command line arguments
call gcmdln()
if (output_file.eq.''.and.euler_output_file.eq.'') then
    call usage('fltinv: output file name is required (usage:general)')
endif
if (inversion_mode.eq.'') then
    call usage('fltinv: inversion mode is required (usage:general)')
endif


! Run the main fltinv subroutines
call read_inputs()      ! In fltinv_io.f90
call calc_disp_gfs()    ! In fltinv_gf.f90
call calc_stress_gfs()  ! In fltinv_gf.f90
call calc_euler_gfs()   ! In fltinv_gf.f90
call run_inversion()    ! In fltinv.f90 (below)
call write_solution()   ! In fltinv_io.f90

end program




!--------------------------------------------------------------------------------------------------!
!--------------------------------------------------------------------------------------------------!
!----------------------------------- INVERSION SUBROUTINES ----------------------------------------!
!--------------------------------------------------------------------------------------------------!
!--------------------------------------------------------------------------------------------------!

subroutine run_inversion()
!----
! Run the selected inversion algorithm. Much like the main fltinv program, this subroutine is
! essentially a wrapper for starting an inversion subroutine.
!----

use io, only: stdout, stderr, verbosity
use solver, only: load_array, load_constraints

use fltinv, only: inversion_mode, &
                  euler_file

implicit none


if (verbosity.ge.1) then
    write(stdout,*) 'run_inversion: starting'
endif


! Run the selected inversion algorithm
if (inversion_mode.eq.'lsqr') then
    call invert_lsqr()                         ! In fltinv_lsqr.f90
elseif (inversion_mode.eq.'anneal') then
    call invert_anneal()                       ! In fltinv_anneal.f90
elseif (inversion_mode.eq.'anneal-psc') then
    if (euler_file.ne.'none') then
        call invert_anneal_euler_psc()         ! In fltinv_anneal_euler_psc.f90
    else
        call invert_anneal_psc()               ! In fltinv_anneal_psc.f90
    endif
else
    write(stderr,*) 'run_inversion: no inversion mode named '//trim(inversion_mode)
    write(stderr,*) 'Options for inversion mode:'
    write(stderr,*) '    lsqr'
    write(stderr,*) '    anneal'
    call usage(     '    anneal-psc (usage:general)')
endif

if (verbosity.ge.1) then
    write(stdout,*) 'run_inversion: finished'
    write(stdout,*)
endif

return
end subroutine run_inversion
